src/compile/data/filterinvalid.ts

```diff
@@ -0,0 +1,68 @@
+import {isScaleChannel} from '../../channel';
+import {field as fieldRef, FieldDef} from '../../fielddef';
+import {hasContinuousDomain, ScaleType} from '../../scale';
+import {Dict, extend, keys} from '../../util';
+import {VgTransform} from '../../vega.schema';
+import {ModelWithField} from '../model';
+import {DataFlowNode} from './dataflow';
+
+export class FilterInvalidNode extends DataFlowNode {
+  private fieldDefs: Dict<FieldDef<string>>;
+
+  public clone() {
+    return new FilterInvalidNode(extend({}, this.fieldDefs));
+  }
+
+  constructor(fieldDefs: Dict<FieldDef<string>>) {
+   super();
+
+   this.fieldDefs = fieldDefs;
+  }
+
+  public static make(model: ModelWithField): FilterInvalidNode {
+
+    const filter = model.reduceFieldDef((aggregator: Dict<FieldDef<string>>, fieldDef, channel) => {
+      const scaleComponent = isScaleChannel(channel) && model.getScaleComponent(channel);
+      if (scaleComponent) {
+        const scaleType = scaleComponent.get('type');
+
+        // only automatically filter null for continuous domain since discrete domain scales can handle invalid values.
+        if (hasContinuousDomain(scaleType) && !fieldDef.aggregate) {
+          aggregator[fieldDef.field] = fieldDef;
+        }
+      }
+      return aggregator;
+    }, {} as Dict<FieldDef<string>>);
+
+    if (!keys(filter).length) {
+      return null;
+    }
+
+  return new FilterInvalidNode(filter);
+  }
+
+  get filter() {
+    return this.fieldDefs;
+  }
+
+  // create the VgTransforms for each of the filtered fields
+  public assemble(): VgTransform[] {
+
+     return keys(this.filter).reduce((vegaFilters, field) => {
```

Comment:
The old assemble in nullfilter.ts produces one filter per `FilterInvalidNode` (`NullFilterNode`), which is more efficient as it requires only one pass over the whole dataset.   We should do the same here. 

For example, we shouldn't see the diff in examples/compiled/trellis_row_column.vg.json.